### PR TITLE
[no squash] Optimize JSON string (de)serialization routines

### LINKED ISSUE #12250: Improve performance of `deSerializeJsonString(IfNeeded)`
## Problem

`deSerializeJsonStringIfNeeded` shows up frequently in profiling during moments of stutter (emscripten build). Within this function, most of the time is spent inside std::stringstream methods.

Relevant code:

https://github.com/minetest/minetest/blob/ec4a789b4f56d7918b6c964e59b9903af5e17fcd/src/util/serialize.cpp#L249-L288

The main issue seems to be that std::stringstream has pathologically bad performance when reading/writing single characters. I'm not aware of the reasons this is the case.

Here's a benchmark to demonstrate the problem: https://gist.github.com/paradust7/a7597d92b4f2cdca0f54dd7ebc09b6fc

Benchmark results on Linux X86-64:

```
$ g++ -O3 benchmark.cpp -o benchmark
$ ./benchmark
use_string_stream with size = 1000000 and loops = 10 took 107 milliseconds
use_vector_char with size = 1000000 and loops = 10 took 13 milliseconds
use_string_stream with size = 10 and loops = 1000000 took 299 milliseconds
use_vector_char with size = 10 and loops = 1000000 took 89 milliseconds
```

Benchmark results on Emscripten running on Node:
```
$ em++ -O3 benchmark.cpp -o benchmark.js -sINITIAL_MEMORY=1000mb
$ node ./benchmark.js
use_string_stream with size = 1000000 and loops = 10 took 273 milliseconds
use_vector_char with size = 1000000 and loops = 10 took 19 milliseconds
use_string_stream with size = 10 and loops = 1000000 took 355 milliseconds
use_vector_char with size = 10 and loops = 1000000 took 97 milliseconds
```

## Solutions

I am investigating a solution. This issue is a placeholder to keep track of the discussion.

### PR DESCRIPTION
**Pull request by sfan5**
closes #12250
benchmark results https://github.com/minetest/minetest/issues/12250#issuecomment-1114608456 (regressions were fixed)

## To do

This PR is Ready for Review.

## How to test

Run the unittests.

### CODE REVIEW

File: src/util/serialize.cpp [hunk 1]
```diff
@@ -120,121 +119,156 @@ std::string deSerializeString32(std::istream &is)
 }
 
 ////
-//// JSON
+//// JSON-like strings
 ////
 
+const char hex_chars[] = "0123456789abcdef";
```
Comment (paradust7) [inline src/util/serialize.cpp:125]: See util/hex.h

File: src/util/serialize.cpp [hunk 2]
```diff
@@ -120,121 +120,152 @@ std::string deSerializeString32(std::istream &is)
...
+				memcpy(hexdigits, &s[i], 4);
+				i += 4;
+				hexdigits[4] = 0;
+
+				std::istringstream tmp_is(hexdigits, std::ios::binary);
+				tmp_is >> std::hex >> hexnumber;
+				// Note that this does not work for anything other than ASCII
+				// but these functions do not actually interact with real JSON input.
+				s[w++] = (char) hexnumber;
+				break;
```
Comment (paradust7) [inline src/util/serialize.cpp:227]: util/hex.h is missing a good interface for doing this. But for now could you do: (haven't tested this)

```suggestion
				u16 hexnumber = 0;
				for (int k = 0; k < 4; k++) {
					if (i >= len)
						throw SerializationError("JSON string ended prematurely");
					unsigned char tmpchar;
					if (!hex_digit_decode(s[i++], tmpchar))
						throw SerializationError("JSON string invalid unicode");
					hexnumber <<= 4;
					hexnumber |= tmpchar;
				}
				s[w++] = (char) hexnumber;
``````

Comment (sfan5) [reply src/util/serialize.cpp:227]: I think I prefer looks over performance here :wink:

Comment (sfan5) [reply src/util/serialize.cpp:227]: Actually, this both looks good and will be faster:
```cpp
// length check
unsigned char c[4] = {};
for (int j = 0; i < 4; i++))
    hex_digit_decode(s[i+j], &c[j]);
u32 hexnumber = c[3] << 12 | c[2] << 8 | c[1] << 4 | c[0];
// the rest
```

Comment (paradust7) [reply src/util/serialize.cpp:227]: You and me have a very different aesthetic sense! I avoid `stringstream` like the plague. Ancient C++ IO APIs are wildly inefficient.

But yes, that is much better. Note that the two i's on the for line should be j.

Comment (paradust7) [reply src/util/serialize.cpp:227]: The order of the hex digits also needs to be reversed. c[0] will have the most-significant digit.

**Comment by paradust7** [comment]
@sfan5 I'm confused about your force push to quicker, it seems to have reverted a bunch of changes.

**Comment by sfan5** [comment]
Which ones? I am working on this on the same branch so nothing unintentional should have come in.

File: src/util/serialize.cpp (hunk 2 above)
Comment (sfan5) [reply src/util/serialize.cpp]: Yes. I stared at it for about 15 minutes because I made three mistakes quickly writing this down...

**Comment by paradust7** [comment]
serializeJsonString went back to using append rather than push_back

**Comment by sfan5** [comment]
Ah, I never changed that.
You're suggesting
```cpp
tmp.push_back('\\');
tmp.push_back('f');
```
over
```cpp
tmp.append("\\f");
```
right?

**Comment by paradust7** [comment]
Oh, ok. I got confused. I thought you had included that change.

It's much easier to keep track with a pull request and a sequence of commits. Is there a reason not to do that?

**Comment by paradust7** [comment]
Or was the force push just to rebase? Maybe I totally misinterpreted this.

**Comment by sfan5** [comment]
Right now I am keeping the PR commits in a "mergeable" form (just two with useful descriptions), that's why I keep force pushing. You're right that I could do this just before merge instead of during development.

File: src/util/serialize.cpp [hunk 3]
```diff
@@ -120,121 +119,148 @@ std::string deSerializeString32(std::istream &is)
...
 				break;
 			default: {
 				if (c >= 32 && c <= 126) {
-					os << c;
+					tmp.push_back(c);
 				} else {
-					u32 cnum = (u8)c;
-					os << "\\u" << std::hex << std::setw(4)
-						<< std::setfill('0') << cnum;
+					// We pretend that Unicode codepoints map to bytes (they don't)
```
Comment (appgurueu) [inline src/util/serialize.cpp:159]: Does this mean Minetest's JSON serialization has always been severely broken regarding bytestrings?

File: src/util/serialize.cpp [hunk 4]
```diff
@@ -120,121 +119,148 @@ std::string deSerializeString32(std::istream &is)
...
+	assert(s.size() >= 2);
+	assert(s.front() == '"' && s.back() == '"');
+
+	size_t w = 0; // write index
+	size_t i = 1; // read index
+	const size_t len = s.size() - 1; // string length with trailing quote removed
+
+	while (i < len) {
+		char c = s[i++];
+		assert(c != '"');
```
Comment (SmallJoker) [inline src/util/serialize.cpp:185]: What about escaped quotation marks within the string?

Comment (sfan5) [reply src/util/serialize.cpp:185]: They're handled below and skipped over (this line does not run for every character in the input)

**Review by SmallJoker — APPROVED** [review]
Works. I trust @paradust7's benchmarks about the performance (and the impressive speed-up).

**Comment by runsy** [comment]
Maye related?

[](https://github.com/runsy/petz/issues/105)

**Comment by runsy** [comment]
No, cos if I check for nil another error:

> AsyncErr: Lua: Runtime error from mod 'petz' in callback luaentity_Activate(): /opt/minetest/bin/../builtin/common/serialize.lua:191: bad argument #1 to 'loadstring' (string expected, got table)
> stack traceback:
> 	[C]: in function 'loadstring'
> 	/opt/minetest/bin/../builtin/common/serialize.lua:191: in function 'deserialize'
> 	...in/../games/juanchi/mods/petz/petz/api/api_init_prop.lua:294: in function 'set_initial_properties'
> 	...bin/../games/juanchi/mods/petz/petz/petz/pony_mobkit.lua:105: in function <...bin/../games/juanchi/mods/petz/petz/petz/pony_mobkit.lua:103>

**Pull request merged**